`round()` - New CSS functional notation (#22212)
mdn/content: files/en-us/web/css/round/index.md

---
title: round()
slug: Web/CSS/round
page-type: css-function
tags:
  - CSS
  - CSS Function
  - Function
  - Math
  - Reference
  - Web
  - round
  - Experimental
browser-compat: css.types.round
---

{{CSSRef}}{{SeeCompatTable}}

> The **`round()`** [CSS](/en-US/docs/Web/CSS) [function](/en-US/docs/Web/CSS/CSS_Functions) is a stepped-value function that returns a rounded number based on a selected rounding strategy.

Reviewer: I infer that a stepped value means that it returns values in increments. But it isn't a term in general use, and I think on the criteria "a middle school child should understand this", they will get just as much from the following:

```suggestion
The **`round()`** [CSS](/en-US/docs/Web/CSS) [function](/en-US/docs/Web/CSS/CSS_Functions) returns a rounded number based on a selected rounding strategy.
```

## Syntax

```css
width: round(var(--width), 50px);
width: round(up, 101, 10);
width: round(down, 106, 10);
margin: round(to-zero, -105, 10);
```

### Parameter

The `round(<rounding-strategy>?, A, B)` function contains an optional rounding strategy, and two calculations `A` and `B` as its parameter.

> - `<rounding-strategy>`
>   - : `up` - Choose upper `B`. Equivalent to JS `Math.ceil()`.
>   - : `down` - Choose lower `B`. Equivalent to JS `Math.floor()`.
>   - : `nearest` - (default) Rounding to the nearest integer. Equivalent to JS `Math.round()`.
>   - : `to-zero` - Choose whichever of lower `B` and upper `B` that has the smallest absolute difference from `0`. Equivalent to JS `Math.trunc()`.
> 
> - `A`
>   - : A calculations that resolves to {{CSSxREF("&lt;number&gt;")}}, {{CSSxREF("&lt;dimension&gt;")}}, or {{CSSxREF("&lt;percentage&gt;")}}.
> 
> - `B`
>   - : A calculations that resolves to {{CSSxREF("&lt;number&gt;")}}, {{CSSxREF("&lt;dimension&gt;")}}, or {{CSSxREF("&lt;percentage&gt;")}}.

Reviewer: It is a limitation of the markdown that you can't have multiple definitions for a term. This modification is required to make it render properly. I'll merge this, because there is no way around that. 

```suggestion
- `<rounding-strategy>`

  - : The rounding strategy.
     This may be one of the following values:

     - `up`
       - : Choose upper `B`. Equivalent to JS `Math.ceil()`.
     - `down`
       - : Choose lower `B`. Equivalent to JS `Math.floor()`.
     - `nearest`
       - : (default) Rounding to the nearest integer. Equivalent to JS `Math.round()`.
     - `to-zero`
       - : Choose whichever of lower `B` and upper `B` that has the smallest absolute difference from `0`. Equivalent to JS `Math.trunc()`.

- `A`
  - : A calculations that resolves to {{CSSxREF("&lt;number&gt;")}}, {{CSSxREF("&lt;dimension&gt;")}}, or {{CSSxREF("&lt;percentage&gt;")}}.

- `B`
  - : A calculations that resolves to {{CSSxREF("&lt;number&gt;")}}, {{CSSxREF("&lt;dimension&gt;")}}, or {{CSSxREF("&lt;percentage&gt;")}}.
```

### Return value

The value of `A`, rounded according to the rounding strategy, to the nearest integer multiple of `B` either above or below `A`.

- If `B` is 0, the result is `NaN`.
- If `A` and `B` are both `infinite`, the result is `NaN`.
- If `A` is infinite but `B` is finite, the result is the same `infinity`.

> - If `A` is finite but `B` is infinite, the result depends on the rounding strategy and the sign of `A`:
>   - : `up` - If `A` is positive (not zero), return `+∞`. If `A` is `0⁺`, return `0⁺`. Otherwise, return `0⁻`.
>   - : `down` - If `A` is negative (not zero), return `−∞`. If `A` is `0⁻`, return `0⁻`. Otherwise, return `0⁺`.
>   - : `nearest`, `to-zero` - If `A` is positive or `0⁺`, return `0⁺`. Otherwise, return `0⁻`.

Reviewer: Same thing - the definition list can't work like this. But in this case I don't think we need to make it a definition list, so have made a plain list. If you want it to be a definition list, then you'd have to modify it as I did above. (feel free if you want to)

```suggestion
- If `A` is finite but `B` is infinite, the result depends on the rounding strategy and the sign of `A`:

  - `up` - If `A` is positive (not zero), return `+∞`. If `A` is `0⁺`, return `0⁺`. Otherwise, return `0⁻`.
  - `down` - If `A` is negative (not zero), return `−∞`. If `A` is `0⁻`, return `0⁻`. Otherwise, return `0⁺`.
  - `nearest`, `to-zero` - If `A` is positive or `0⁺`, return `0⁺`. Otherwise, return `0⁻`.
```

---
title: round()
slug: Web/CSS/round
page-type: css-function
tags:
  - CSS
  - CSS Function
  - Function
  - Math
  - Reference
  - Web
  - round
  - Experimental
browser-compat: css.types.round
---

{{CSSRef}}{{SeeCompatTable}}

The **`round()`** [CSS](/en-US/docs/Web/CSS) [function](/en-US/docs/Web/CSS/CSS_Functions) returns a rounded number based on a selected rounding strategy.

## Syntax

```css
width: round(var(--width), 50px);
width: round(up, 101, 10);
width: round(down, 106, 10);
margin: round(to-zero, -105, 10);
```

### Parameter

> The `round(<rounding-strategy>?, A, B)` function contains an optional rounding strategy, and two calculations `A` and `B` as its parameter.

Reviewer: I understand that the ? after `<rounding-strategy>` means "optionally one of these strategies". What will happen if you don't have a strategy listed? I _guess_ its a case where the value is in theory optional, but you can't use the function without it?

When you say "two calculations `A` and `B`" do you mean that mathematical expression?

Reviewer: OK, starting to understand this after reading the return value section. My understanding is that "A" is the value to be rounded, and "B" is the rounding interval, and the rounding strategy determines whether you round up, down, to nearest integer, and so on. Is that right?

So if the value of the rounding strategy is not specified you'd use `nearest`. B is not used in this case - right?

Are the names "A" and "B" forced upon you from somewhere? If not, we should give them useful names:

```suggestion
The `round(<rounding-strategy>, valueToRound, roundingInterval)` function specifies an optional rounding strategy, a value (or mathematical expression) to be rounded and a rounding interval (or mathematical expression).
The `valueToRound` is rounded according to the rounding strategy, to the nearest integer multiple of `roundingInterval`.
```

- `<rounding-strategy>`

  - : The rounding strategy.
     This may be one of the following values:

     - `up`

>        - : Choose upper `B`. Equivalent to JS `Math.ceil()`.

Reviewer: I don't understand what you mean by "Choose upper `B` ". 

In your real example all I can infer from this is that rounding has something to do with "10", and that 101 is meaningless. 
```
width: round(up, 101, 10);
```
Effectively I have the same question for `down`

>        - : Choose upper `B`. Equivalent to JS `Math.ceil()`.

Reviewer: ```suggestion
       - : Round `valueToRound` to the next more positive `roundingInterval` value. This is equivalent to the JavaScript `Math.ceil()`.
```

Would be good to link to the `Math.ceil()` javascript docs too

     - `down`

>        - : Choose lower `B`. Equivalent to JS `Math.floor()`.

Reviewer: ```suggestion
       - : Round `valueToRound` down to the nearest integer multiple of `roundingInterval` (if the value is negative, it will become  "more negative"). This is equivalent to the JavaScript `Math.floor()` method.
```

     - `nearest`

>        - : (default) Rounding to the nearest integer. Equivalent to JS `Math.round()`.

Reviewer: This I get. Does that mean A, B are ignored if this is specified?

>        - : (default) Rounding to the nearest integer. Equivalent to JS `Math.round()`.

Reviewer: Do you mean this:

```suggestion
       - : (default) Round `valueToRound` to the nearest integer (`roundingInterval` is ignored). Equivalent to JS `Math.round()`.
```

What if you're on 0.5 - do you round up or down?

     - `to-zero`

>        - : Choose whichever of lower `B` and upper `B` that has the smallest absolute difference from `0`. Equivalent to JS `Math.trunc()`.

Reviewer: I don't know what you mean by lower B and upper B. Can you work through a real example?

>        - : Choose whichever of lower `B` and upper `B` that has the smallest absolute difference from `0`. Equivalent to JS `Math.trunc()`.

Reviewer: ```suggestion
       - : Round `valueToRound` to the nearest integer multiple of `roundingInterval` closer to/towards zero (a positive number will decrease, while a negative value will become "less negative"). This is equivalent to the JavaScript `Math.trunc()` method.
```

> - `A`

Reviewer: ```suggestion
- `roundingValue`
```

>   - : A calculations that resolves to {{CSSxREF("&lt;number&gt;")}}, {{CSSxREF("&lt;dimension&gt;")}}, or {{CSSxREF("&lt;percentage&gt;")}}.

Reviewer: ```suggestion
  - : A calculation that resolves to {{CSSxREF("&lt;number&gt;")}}, {{CSSxREF("&lt;dimension&gt;")}}, or {{CSSxREF("&lt;percentage&gt;")}}.
```

> - `B`

Reviewer: ```suggestion
- `roundingInterval`
```

>   - : A calculations that resolves to {{CSSxREF("&lt;number&gt;")}}, {{CSSxREF("&lt;dimension&gt;")}}, or {{CSSxREF("&lt;percentage&gt;")}}.

Reviewer: ```suggestion
  - : A {{CSSxREF("&lt;number&gt;")}}, {{CSSxREF("&lt;dimension&gt;")}}, or {{CSSxREF("&lt;percentage&gt;")}}, or a mathematical expression that resolves to to one of those values.
```

### Return value

> The value of `A`, rounded according to the rounding strategy, to the nearest integer multiple of `B` either above or below `A`.

Reviewer: ```suggestion
The value of `roundingValue`, rounded to the nearest lower or higher integer multiple of `roundingInterval`, depending on the  `rounding strategy`
```

> - If `B` is 0, the result is `NaN`.
> - If `A` and `B` are both `infinite`, the result is `NaN`.
> - If `A` is infinite but `B` is finite, the result is the same `infinity`.
> - If `A` is finite but `B` is infinite, the result depends on the rounding strategy and the sign of `A`:
> 
>   - `up` - If `A` is positive (not zero), return `+∞`. If `A` is `0⁺`, return `0⁺`. Otherwise, return `0⁻`.
>   - `down` - If `A` is negative (not zero), return `−∞`. If `A` is `0⁻`, return `0⁻`. Otherwise, return `0⁺`.
>   - `nearest`, `to-zero` - If `A` is positive or `0⁺`, return `0⁺`. Otherwise, return `0⁻`.
> - The argument calculations can resolve to {{CSSxREF("&lt;number&gt;")}}, {{CSSxREF("&lt;dimension&gt;")}}, or {{CSSxREF("&lt;percentage&gt;")}}, but must have the same type, or else the function is invalid; the result will have the same type as the arguments.
> - If `A` is exactly equal to an integer multiple of `B`, `round()` resolves to `A` exactly (preserving whether `A` is `0⁻` or `0⁺`, if relevant). Otherwise, there are two integer multiples of `B` that are potentially "closest" to `A`, lower `B` which is closer to `−∞` and upper `B` which is closer to `+∞`.

Reviewer: ```suggestion
- If `roundingInterval` is 0, the result is `NaN`.
- If `valueToRound` and `roundingInterval` are both `infinite`, the result is `NaN`.
- If `valueToRound` is infinite but `roundingInterval` is finite, the result is the same `infinity`.
- If `valueToRound` is finite but `roundingInterval` is infinite, the result depends on the rounding strategy and the sign of `A`:

  - `up` - If `valueToRound` is positive (not zero), return `+∞`. If `valueToRound` is `0⁺`, return `0⁺`. Otherwise, return `0⁻`.
  - `down` - If `valueToRound` is negative (not zero), return `−∞`. If `valueToRound` is `0⁻`, return `0⁻`. Otherwise, return `0⁺`.
  - `nearest`, `to-zero` - If `valueToRound` is positive or `0⁺`, return `0⁺`. Otherwise, return `0⁻`.
- The argument calculations can resolve to {{CSSxREF("&lt;number&gt;")}}, {{CSSxREF("&lt;dimension&gt;")}}, or {{CSSxREF("&lt;percentage&gt;")}}, but must have the same type, or else the function is invalid; the result will have the same type as the arguments.
- If `valueToRound` is exactly equal to an integer multiple of `roundingInterval`, `round()` resolves to `valueToRound` exactly (preserving whether `valueToRound` is `0⁻` or `0⁺`, if relevant). Otherwise, there are two integer multiples of `roundingInterval` that are potentially "closest" to `valueToRound`, lower `roundingInterval` which is closer to `−∞` and upper `roundingInterval` which is closer to `+∞`.
```

### Formal syntax

{{CSSSyntax}}

> 

Reviewer: Is it possible to add a live example or interactive example showing this being used in practice?

Reviewer: @ramiy Can you please estimate whe you might be able to add this example?

Author: @hamishwillee Can we merge without examples for now? I will create a separate PR when I have some spare time.

Reviewer: Yes. I may create that separate PR, time permitting.

## Syntax

```css
width: round(var(--width), 50px);
width: round(up, 101, 10);
width: round(down, 106, 10);
margin: round(to-zero, -105, 10);
```

### Parameter

The `round(<rounding-strategy>, valueToRound, roundingInterval)` function specifies an optional rounding strategy, a value (or mathematical expression) to be rounded and a rounding interval (or mathematical expression).
The `valueToRound` is rounded according to the rounding strategy, to the nearest integer multiple of `roundingInterval`.

- `<rounding-strategy>`

  - : The rounding strategy.
    This may be one of the following values:

    - `up`
      - : Round `valueToRound` up to the nearest integer multiple of `roundingInterval` (if the value is negative, it will become "more positive"). This is equivalent to the JavaScript `Math.ceil()` method.
    - `down`
      - : Round `valueToRound` down to the nearest integer multiple of `roundingInterval` (if the value is negative, it will become "more negative"). This is equivalent to the JavaScript `Math.floor()` method.
    - `nearest` (default)

>       - : Round `valueToRound` to the nearest integer (`roundingInterval` is ignored). Equivalent to JS `Math.round()`.

Reviewer: Note, use "JavaScript" not "JS"

```suggestion
      - : Round `valueToRound` to the nearest integer (`roundingInterval` is ignored). Equivalent to JavaScript [`Math.round()`](/en-US/docs/Web/JavaScript/Reference/Global_Objects/Math/round).
```

>        - : (default) Round `valueToRound` to the nearest integer (`roundingInterval` is ignored). Equivalent to JS `Math.round()`.

Reviewer: What if the value is in the middle - i.e. 0.5? Which way will it round?

>       - : Round `valueToRound` to the nearest integer (`roundingInterval` is ignored). Equivalent to JavaScript [`Math.round()`](/en-US/docs/Web/JavaScript/Reference/Global_Objects/Math/round).

Reviewer: Fixing. If half way though it rounds up, according to spec.
```suggestion
      - : Round `valueToRound` to the nearest integer multiple of `roundingInterval`, which may be either above or below the value.
        If the `valueToRound` is equally close to both rounding targets, it will be rounded up.
        Equivalent to JavaScript [`Math.round()`](/en-US/docs/Web/JavaScript/Reference/Global_Objects/Math/round).
```

>     - `up`
>       - : Round `valueToRound` up to the nearest integer multiple of `roundingInterval` (if the value is negative, it will become "more positive"). This is equivalent to the JavaScript `Math.ceil()` method.
>     - `down`
>       - : Round `valueToRound` down to the nearest integer multiple of `roundingInterval` (if the value is negative, it will become "more negative"). This is equivalent to the JavaScript `Math.floor()` method.
>     - `nearest` (default)
>       - : Round `valueToRound` to the nearest integer (`roundingInterval` is ignored). Equivalent to JavaScript [`Math.round()`](/en-US/docs/Web/JavaScript/Reference/Global_Objects/Math/round).
>     - `to-zero`
>       - : Round `valueToRound` to the nearest integer multiple of `roundingInterval` closer to/towards zero (a positive number will decrease, while a negative value will become "less negative"). This is equivalent to the JavaScript `Math.trunc()` method.

Reviewer: ```suggestion
    - `up`
      - : Round `valueToRound` up to the nearest integer multiple of `roundingInterval` (if the value is negative, it will become "more positive"). This is equivalent to the JavaScript [`Math.ceil()`](/en-US/docs/Web/JavaScript/Reference/Global_Objects/Math/ceil) method.
    - `down`
      - : Round `valueToRound` down to the nearest integer multiple of `roundingInterval` (if the value is negative, it will become "more negative"). This is equivalent to the JavaScript [`Math.floor()`](/en-US/docs/Web/JavaScript/Reference/Global_Objects/Math/floor) method.
    - `nearest` (default)
      - : Round `valueToRound` to the nearest integer (`roundingInterval` is ignored). Equivalent to JavaScript [`Math.round()`](/en-US/docs/Web/JavaScript/Reference/Global_Objects/Math/round).
    - `to-zero`
      - : Round `valueToRound` to the nearest integer multiple of `roundingInterval` closer to/towards zero (a positive number will decrease, while a negative value will become "less negative"). This is equivalent to the JavaScript [`Math.trunc()`](/en-US/docs/Web/JavaScript/Reference/Global_Objects/Math/trunc) method.
```

- `roundingInterval`
  - : The rounding interval.
    This is a {{CSSxREF("&lt;number&gt;")}}, {{CSSxREF("&lt;dimension&gt;")}}, or {{CSSxREF("&lt;percentage&gt;")}}, or a mathematical expression that resolves to to one of those values.

### Return value

The value of `valueToRound`, rounded to the nearest lower or higher integer multiple of `roundingInterval`, depending on the `rounding strategy`.

- If `roundingInterval` is 0, the result is `NaN`.
- If `valueToRound` and `roundingInterval` are both `infinite`, the result is `NaN`.
- If `valueToRound` is infinite but `roundingInterval` is finite, the result is the same `infinity`.
- If `valueToRound` is finite but `roundingInterval` is infinite, the result depends on the rounding strategy and the sign of `A`:

  - `up` - If `valueToRound` is positive (not zero), return `+∞`. If `valueToRound` is `0⁺`, return `0⁺`. Otherwise, return `0⁻`.
  - `down` - If `valueToRound` is negative (not zero), return `−∞`. If `valueToRound` is `0⁻`, return `0⁻`. Otherwise, return `0⁺`.

>   - `nearest`, `to-zero` - If `valueToRound` is positive or `0⁺`, return `0⁺`. Otherwise, return `0⁻`.

Reviewer: What if the value is exactly half way through the interval? i.e. the result could be nearest up OR nearest down. How is that resolved?

Author: No idea...
